File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/KeyfileImportViewModel.swift
import Foundation
import Observation
import CryptoKit
import Flagship
import FlagshipAPI
import FlagshipCore

/// Drives "Import backup file" — bring this device into an account using
/// its `.flagshipkey` backup. The flow mirrors the credentialed takeover
/// in `RealAccountLoginViewModel`, swapping the WebAuthn-PRF unwrap for a
/// passphrase-decrypt of the keyfile:
///
///   1. The host hands us the picked file's text + the user's passphrase.
///   2. `Keyfile.unwrap` decrypts the UMK seed (argon2id + AES-256-GCM).
///   3. Point the Keystore at this account's slot + `installUMK`.
///   4. Initiate the takeover re-pair (new IRK derives from the installed
///      UMK), then complete it. On `.finalized` the host flips AppState.
///
/// Errors map to the approved copy: wrong passphrase → "That passphrase
/// didn't open the file."; not a keyfile → "This isn't a Flagship key
/// file."
@MainActor
@Observable
public final class KeyfileImportViewModel {
    public enum Phase: Equatable, Sendable {
        case idle
        case working
        /// Re-pair initiated — the takeover grace is running server-side.
        case completed(username: String, completesAt: Int64)
        /// Re-pair complete — the host flips AppState to paired on this.
        case finalized(username: String)
        case failed(String)
    }

    public private(set) var phase: Phase = .idle
    public var passphrase: String = ""

    private let server: any FlagshipServerClient
    /// Seam so tests can assert the imported UMK is installed without
    /// piercing the Secure Enclave. Defaults to the real Keystore write.
    private let installUMK: @MainActor (SymmetricKey, String) async throws -> Void

    public init(
        server: any FlagshipServerClient,
        installUMK: @escaping @MainActor (SymmetricKey, String) async throws -> Void = { seed, reason in
            try await Keystore.installUMK(seed, reason: reason)
        }
    ) {
        self.server = server
        self.installUMK = installUMK
    }

    public var canImport: Bool {
        !passphrase.isEmpty
    }

    /// Unwrap the keyfile, install the UMK, and initiate the takeover
    /// re-pair. `fileText` is the raw `.flagshipkey` contents the host
    /// read from the picked file.
    public func importBackup(fileText: String) async {
        guard canImport else {
            phase = .failed("Enter the passphrase for this backup file.")
            return
        }
        phase = .working

        // 1 — Decrypt the keyfile.
        let seed: SymmetricKey
        let meta: Keyfile.Meta
        do {
            (seed, meta) = try Keyfile.unwrap(fileText: fileText, passphrase: passphrase)
        } catch let e as Keyfile.KeyfileError {
            phase = .failed(humanizedKeyfileError(e))
            return
        } catch {
            phase = .failed("This isn't a Flagship key file.")
            return
        }

        let username = meta.username

        // 2 — Install the recovered UMK into THIS account's slot, so a
        // second imported cloud lands in its own slot and never clobbers
        // an already-present profile. installUMK resets the IRK lineage
        // to v1 under the imported UMK.
        Keystore.setActiveProfile(username)
        do {
            try await installUMK(seed, "Bring this device into your Flagship account")
        } catch {
            phase = .failed("Couldn't install your account key. \(HumanError.humanize(error))")
            return
        }

        // 3 — Initiate the takeover re-pair. New IRK derives from the
        // just-installed UMK. On a fresh device we don't hold the
        // displaced key, so the old-pubkey slot carries the new pubkey
        // (the Worker keys the takeover on the username row).
        do {
            let resp = try await initiateTakeoverRePair(username: username)
            phase = .completed(username: username, completesAt: resp.completesAt)
        } catch ScreensClientError.http(let status, let message) where status == 401 && (message ?? "").contains("totpProof") {
            // #52 — the account has a second factor enrolled, which the
            // cloud now requires at initiate even for single-device
            // accounts. The keyfile sheet has no second-factor field
            // (yet); route the user to the sign-in flow, which prompts
            // for it.
            phase = .failed("This account has a second factor enrolled. Use \"I already have an account\" to sign in — it will ask for your authenticator or recovery code.")
        } catch {
            phase = .failed("Couldn't start bringing this device in. \(HumanError.humanize(error))")
        }
    }

    /// Finalize the takeover once its grace has elapsed. The complete
    /// endpoint is a public, idempotent CAS-swap. 404 == already
    /// swapped (success); 425 == too early (stay in grace); 403/409 ==
    /// objected (cancelled).
    public func completeTakeover() async {
        guard case .completed(let username, let completesAt) = phase else { return }
        phase = .working
        do {
            _ = try await server.completeRePair(username: username)
            phase = .finalized(username: username)
        } catch ScreensClientError.http(let status, _) where status == 404 {
            phase = .finalized(username: username)
        } catch ScreensClientError.http(let status, _) where status == 425 {
            phase = .completed(username: username, completesAt: completesAt)
        } catch ScreensClientError.http(let status, _) where status == 403 || status == 409 {
            phase = .failed("This was cancelled. If it's still you, try again.")
        } catch ScreensClientError.http(let status, _) where status == 410 {
            // #52 — completion window elapsed; the cloud swept the row.
            phase = .failed("This expired before it was completed. Start again.")
        } catch {
            phase = .failed("Couldn't finish bringing this device in. \(HumanError.humanize(error))")
        }
    }

    private func initiateTakeoverRePair(username: String) async throws -> RePairInitiateResponse {
        let newKey = try await Keystore.deriveIRK(reason: "Authorize this device")
        let newPubHex = HexUtil.encode(newKey.publicKey.rawRepresentation)
        let oldPubHex = newPubHex
        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let canonical = RePairInitiate.canonicalBytes(
            username: username,
            newIrkPubHex: newPubHex,
            oldIrkPubHex: oldPubHex,
            issuedAt: issuedAt
        )
        let signature = try newKey.signature(for: canonical)
        return try await server.initiateRePair(
            username: username,
            body: RePairInitiateRequest(
                request: .init(
                    username: username,
                    newIrkPub: newPubHex,
                    oldIrkPub: oldPubHex,
                    issuedAt: issuedAt
                ),
                signature: HexUtil.encode(signature),
                totpProof: nil
            ),
            ifMatch: nil
        )
    }

    public func reset() {
        phase = .idle
    }

    private func humanizedKeyfileError(_ e: Keyfile.KeyfileError) -> String {
        switch e {
        case .badPassphrase: return "That passphrase didn't open the file."
        case .malformed:     return "This isn't a Flagship key file."
        case .version:       return "This backup was made by a newer version of Flagship. Update the app and try again."
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/KeyfileImportViewModel.kt
// "Import backup file" — bring this device into an account using its
// `.flagshipkey` backup. Mirror of the iOS KeyfileImportViewModel; the
// flow swaps the WebAuthn-PRF unwrap of the credentialed takeover for a
// passphrase-decrypt of the keyfile:
//
//   1. The host hands us the picked file's text + the user's passphrase.
//   2. Keyfile.unwrap decrypts the UMK seed (argon2id + AES-256-GCM).
//   3. Point the Keystore at this account's per-profile slot + installUmk.
//   4. Initiate the takeover re-pair (new IRK derives from the installed
//      UMK), then complete it once the grace elapses. On Completed the
//      host flips AppState to paired (mirrors LoginViewModel).
//
// Errors map to the approved copy: wrong passphrase → "That passphrase
// didn't open the file."; not a keyfile → "This isn't a Flagship key
// file."

package com.flagshipserver.app.viewmodels

import androidx.lifecycle.ViewModel
import com.flagshipserver.app.api.FlagshipServerClient
import com.flagshipserver.app.api.RePairInitiateRequest
import com.flagshipserver.app.core.AppState
import com.flagshipserver.app.core.HexUtil
import com.flagshipserver.app.core.HttpException
import com.flagshipserver.app.core.RePairInitiateClaim
import com.flagshipserver.app.keystore.Keyfile
import com.flagshipserver.app.keystore.Keystore
import com.google.crypto.tink.subtle.Ed25519Sign
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

sealed interface KeyfileImportPhase {
    data object Idle : KeyfileImportPhase
    data object Working : KeyfileImportPhase
    /** Re-pair initiated — the takeover grace is running server-side. */
    data class Grace(val username: String, val completesAt: Long) : KeyfileImportPhase
    /** Re-pair complete — the account is open as a fresh `admin` device. */
    data class Opened(val username: String) : KeyfileImportPhase
    data class Failed(val message: String) : KeyfileImportPhase
}

class KeyfileImportViewModel(
    private val server: FlagshipServerClient,
    private val app: AppState,
    /** Wall-clock injectable for deterministic canonical-bytes tests. */
    private val now: () -> Long = { System.currentTimeMillis() },
) : ViewModel() {

    private val _phase = MutableStateFlow<KeyfileImportPhase>(KeyfileImportPhase.Idle)
    val phase: StateFlow<KeyfileImportPhase> = _phase.asStateFlow()

    private val _passphrase = MutableStateFlow("")
    val passphrase: StateFlow<String> = _passphrase.asStateFlow()
    fun setPassphrase(v: String) { _passphrase.value = v }

    val canImport: Boolean
        get() = _passphrase.value.isNotEmpty()

    /**
     * Unwrap the keyfile, install the UMK, and initiate the takeover
     * re-pair. [fileText] is the raw `.flagshipkey` contents the host
     * read from the picked file.
     */
    suspend fun importBackup(fileText: String) {
        if (!canImport) {
            _phase.value = KeyfileImportPhase.Failed("Enter the passphrase for this backup file.")
            return
        }
        _phase.value = KeyfileImportPhase.Working

        // 1 — Decrypt the keyfile.
        val seed: ByteArray
        val username: String
        try {
            val (s, meta) = Keyfile.unwrap(fileText, _passphrase.value)
            seed = s
            username = meta.username
        } catch (e: Keyfile.KeyfileException) {
            _phase.value = KeyfileImportPhase.Failed(humanizedKeyfileError(e))
            return
        } catch (_: Throwable) {
            _phase.value = KeyfileImportPhase.Failed("This isn't a Flagship key file.")
            return
        }

        try {
            // 2 — Point the Keystore at the recovered cloud's per-profile
            //     device-key slot BEFORE installing, so an imported second
            //     cloud lands in its own slot and never clobbers an
            //     already-present profile. installUmk resets the IRK
            //     lineage to v1 under the imported UMK.
            Keystore.setActiveProfile(username)
            Keystore.installUmk(seed)

            // 3 — Initiate the takeover re-pair. Derive OLD (v1) + NEW (v2)
            //     IRK from the just-installed UMK and sign with the NEW IRK,
            //     mirroring the LoginViewModel credentialed takeover. A
            //     keyfile import is single-device proof, so no totpProof.
            val oldVersion = Keystore.currentIrkVersion()
            val newVersion = oldVersion + 1
            val newSign = Keystore.deriveIRK("Bring this device into your Flagship account", newVersion)
            Keystore.deriveIRK("Confirm import", oldVersion)
            val oldPubHex = pubHexForVersion(oldVersion)
            val newPubHex = pubHexForVersion(newVersion)

            val issuedAt = now()
            val canonical = RePairInitiateClaim.canonicalBytes(
                username = username,
                newIrkPubHex = newPubHex,
                oldIrkPubHex = oldPubHex,
                issuedAt = issuedAt,
            )
            val signature = HexUtil.encode(newSign.sign(canonical))

            val resp = server.initiateRePair(
                username = username,
                body = RePairInitiateRequest(
                    request = RePairInitiateRequest.Inner(
                        username = username,
                        newIrkPub = newPubHex,
                        oldIrkPub = oldPubHex,
                        issuedAt = issuedAt,
                    ),
                    signature = signature,
                    totpProof = null,
                ),
                ifMatch = null,
            )
            Keystore.setPendingIrkRotationVersion(newVersion)
            _phase.value = KeyfileImportPhase.Grace(username = username, completesAt = resp.completesAt)
        } catch (t: Throwable) {
            _phase.value = KeyfileImportPhase.Failed(
                if (t is HttpException && t.status == 401 && t.body.contains("totpProof"))
                    // #52 — the account has a second factor enrolled, which
                    // the Worker now requires at initiate even for single-
                    // device accounts. The keyfile sheet has no second-
                    // factor field (yet); route via the sign-in flow.
                    "This account has a second factor enrolled. Use \"I already have an account\" to sign in — it will ask for your authenticator or recovery code."
                else
                    "Couldn't start bringing this device in: ${t.message}",
            )
        }
    }

    /**
     * Finalize the takeover once its grace has elapsed. The complete
     * endpoint is a public, idempotent CAS-swap. On success we activate
     * the staged IRK rotation locally, open the account as the imported
     * user, leaving the device unnamed.
     */
    suspend fun completeImport() {
        val grace = _phase.value as? KeyfileImportPhase.Grace ?: return
        _phase.value = KeyfileImportPhase.Working
        try {
            server.completeRePair(grace.username)
            Keystore.pendingIrkRotationVersion()?.let { pending ->
                Keystore.setCurrentIrkVersion(pending)
                Keystore.setPendingIrkRotationVersion(null)
            }
            // Left unnamed: administrator status is a capability in the
            // device's signed grant, never a locally invented display name.
            app.completeOnboarding(username = grace.username, pods = emptyList())
            _phase.value = KeyfileImportPhase.Opened(grace.username)
        } catch (t: Throwable) {
            _phase.value = KeyfileImportPhase.Failed(
                if (t is HttpException && t.status == 410)
                    // #52 — completion window elapsed; the cloud swept the row.
                    "This expired before it was completed. Start again."
                else
                    "Couldn't finish bringing this device in: ${t.message}",
            )
        }
    }

    fun reset() {
        _phase.value = KeyfileImportPhase.Idle
    }

    private fun pubHexForVersion(version: Int): String {
        val seed = Keystore.requireIrkSeedForVersion(version)
        val pair = Ed25519Sign.KeyPair.newKeyPairFromSeed(seed)
        return HexUtil.encode(pair.publicKey)
    }

    private fun humanizedKeyfileError(e: Keyfile.KeyfileException): String = when (e.code) {
        Keyfile.KeyfileException.Code.BAD_PASSPHRASE -> "That passphrase didn't open the file."
        Keyfile.KeyfileException.Code.MALFORMED -> "This isn't a Flagship key file."
        Keyfile.KeyfileException.Code.VERSION ->
            "This backup was made by a newer version of Flagship. Update the app and try again."
    }
}
